File: tech_int_prac/valid_parentheses.rb
# Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

# An input string is valid if:

# Open brackets must be closed by the same type of brackets.
# Open brackets must be closed in the correct order.
 

# Example 1:

# Input: s = "()"
# Output: true
# Example 2:

# Input: s = "()[]{}"
# Output: true
# Example 3:

# Input: s = "(]"
# Output: false
 

# Constraints:

# 1 <= s.length <= 104
# s consists of parentheses only '()[]{}'


def is_valid(s)
  return true if s.empty?
  valid_pairs = {
    "(" => ")",
    "{" => "}",
    "[" => "]"
  }
  test_stack = []
  s.each_char do |char|
    if valid_pairs.key? (char)
      test_stack << char
    elsif valid_pairs[test_stack.last] == char
      test_stack.pop
    else
      return false
      break
    end
  end
  test_stack.empty?
end

is_valid("]")

File: tech_int_prac/valid_parentheses.js
// Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

// An input string is valid if:

// Open brackets must be closed by the same type of brackets.
// Open brackets must be closed in the correct order.

var isValid = function(s) {
  let validPairs = {
    "[": "]",
    "{": "}",
    "(": ")"
  };
  let testStack = [];
  for (let i = 0; i < s.length; i++) {
    if (s.charAt(i) in validPairs) {
      testStack.push(s.charAt(i));
    } else if (validPairs[testStack[testStack.length - 1]] === s.charAt(i)) {
      testStack.pop();
    } else {
      return false;
    }
  }
  return !testStack.length;
};

console.log(isValid("[]"));
console.log(isValid(""));
console.log(isValid("{}{}"));
console.log(isValid("}{"));
console.log(isValid("[][][][][]["));
